File: solutions/python/transpose/1/transpose.py
from itertools import zip_longest
from typing import Optional, TypeVar, Iterable

A = TypeVar("A")


def rstrip(it: Iterable[Optional[A]]) -> list[Optional[A]]:
    ls = list(it)
    while ls and ls[-1] is None:
        ls.pop()
    return ls


def transpose(text: str) -> str:
    return "\n".join(
        "".join(c or " " for c in rstrip(col))
        for col in zip_longest(*text.splitlines())
    )

File: solutions/haskell/transpose/1/src/Transpose.hs
module Transpose (transpose) where

import Data.List (dropWhileEnd)
import Data.Maybe (fromMaybe, isNothing, listToMaybe)

transpose :: [String] -> [String]
transpose rows = [map (fromMaybe ' ') $ rstrip col | col <- zipLongest rows]

zipLongest :: [[Char]] -> [[Maybe Char]]
zipLongest = go
  where
    go [] = []
    go xss
        | all null xss = []
        | otherwise = map listToMaybe xss : go (map (drop 1) xss)

rstrip :: [Maybe a] -> [Maybe a]
rstrip = dropWhileEnd isNothing
